### ISSUE TITLE
Discussion of feature: Should extension displays the number of comments?

### ISSUE THREAD

**Issue by 1natsu172** [dideler/toggle-youtube-comments#20]
## Unknown how many comments there are…

current, It's unknown whether the number of comments is 0 or more? 🔢 💬 

👇  Example: Display number of comments on toggle button 👇 

[![https://gyazo.com/3155086ca6065a10c8f1851f6f070ac4](https://i.gyazo.com/3155086ca6065a10c8f1851f6f070ac4.png)](https://gyazo.com/3155086ca6065a10c8f1851f6f070ac4)
> Like this

I feel that usability is better by displaying numbers only when there are comments.
But maybe some people don't want this...? 🤔 

What do you think, everyone?

**Comment by dideler** [comment]
This is interesting. What problem(s) does it solve?

One scenario I can think of, is wanting to know how many comments there are, without wanting to view the comments. This could encourage users to read comments more often - I would be more likely to read the comments if I knew there were a handful for example.

But it also adds more content than is strictly necessary to a simple and clean interface, maybe some users will feel like it makes the UI messy.

How would the "HIDE COMMENTS" button look like?

**Comment by 1natsu172** [comment]
This is my feeling as a user. I am bothered by the comments fundamentally.
Sometimes comments on spoils are written in videos with stories ...
Various opinion comments are written in political videos, comments are seen in the eyes before the video ends, imprinting of thought(ideology) happens ...
However, comments are necessary, and I would like to know what kind of comments are there after watch video. :man_technologist: 

I think that we know the type of video by how many comments are there.
Video with only voting buttons used, video with votes and comments, video with lots of comments but few views ... 🤔 
Simply if there are 0 comments we don't need to press the "SHOW COMMENTS" button.
As @dideler says, if we know how many comments there are, it may encourage to read comments.
How many comments are there gives us various information. I feel that the comment is part of the video information.

***

> But it also adds more content than is strictly necessary to a simple and clean interface, maybe some users will feel like it makes the UI messy.

I am concerned about this as well. Is it suitable for this extension that expresses "lightweight extension"?
Also, there may be users who think that displaying the number of comments is annoying.

……This issue has a part related to the issue (#19)… 📑 

This is also a developer's perspective, but as the page jumps extension have to get the number of comments and re-inject the buttons. The process becomes somewhat complicated ...

`Detect page jump → Get number of comments asynchronously → re-inject button`

Probably it will be such a process…we must carefully consider it.

***

> How would the "HIDE COMMENTS" button look like?

I think the "HIDE COMMENTS" button does not need numbers.
Since the number of comments on YouTube's original is displayed  when opening the comment field.

![](https://i.gyazo.com/a5e7d5091e07c084fc69e9a6ee35bc09.png)

I think that it is only the "SHOW COMMENTS" button that may need the number of comments.

**Comment by dideler** [comment]
I'm becoming more keen on this idea while running into videos with zero comments. It would be good to know that a video has no comments before viewing them as it saves a click.

In terms of format, we could do

<img width="176" alt="screen shot 2018-01-07 at 14 57 56" src="https://user-images.githubusercontent.com/497458/34650573-319d6286-f3bb-11e7-9cb5-e5194acb8898.png">

or, as @1natsu172 originally suggested

<img width="187" alt="screen shot 2018-01-07 at 14 59 08" src="https://user-images.githubusercontent.com/497458/34650581-5a1817ce-f3bb-11e7-8952-523b154c735a.png">

The first option is more readable since it reads like a sentence. The second option is more visually consistent between videos, since the "COMMENTS" part of the text wouldn't move around based on the number of comments.

**Comment by 1natsu172** [comment]
Certainly the first option is readable. But it's only suitable for English ....
I just translated Japanese. With the first option, it very unnatural in Japanese native grammar. This problem cause from the difference between Japanese grammar and English grammar.

🚫 This is unnatural Japanese with first option 
![](https://i.gyazo.com/254d15af3a8d6c9a2afbfa9f99be48ad.png)

✅  This is my suggestion, but not unnatural
![](https://i.gyazo.com/434a8e9373fd880c19f48887d9ff2834.png)

Maybe this is not only case in Japanese.
I think similar problems will cause when supporting other languages ​​in the future. When the language grammar is a different from English grammar.

Of course, we can change the output code for each language, but is it maintainable?

**Comment by dideler** [comment]
I agree. Let's go with the second option of having the number of comments appear behind the text.

@1natsu172 is this something you would like to implement?

**Comment by 1natsu172** [comment]
OK! Let's go second option.

This feature is what I began to say.
So, if you don't extreme hurry implement this features, I'll implement. I'm going to develop this feature in my spare time (both OldYouTube & NewYouTube).

**Comment by dideler** [comment]
I'll work on some of the other open issues this weekend and let you have this one.

By the way, since you've been an active contributor, I've added you as a collaborator. This means you'll have push access and no longer have to work on your fork. Instead you can open Pull Requests directly from inside this repository :smiley:.

**Comment by 1natsu172** [comment]
wow! @dideler thanks inviting me at this repository! 😄 

I understand. I start work on develop this issues's feature.

I was thinking about implementation plan. I also considered the YouTube Data API, but this project is OSS and Public Chrome Extension. So it's difficult to keep the API key secret.
Probably we have to dynamically communicate API keys using PaaS(like Heroku). Moreover, there is problem the limit of YouTube Data API daily request…… 🤔 

So, without using the API, I'll simply, parse the number of comments node by regular expression.

<img width="417" alt="2018-01-11t22-07-55" src="https://user-images.githubusercontent.com/7282145/34827299-03bc0600-f71d-11e7-8c73-a642fd18da47.png">

This is necessary to asynchronously wait for the timing to access this node. There are several patterns such as page navigation(page jump) and direct visit to video page…(and more patterns?)

…I'll report here if any problems with implementation occur.

Anyway, Let's keep up good work 💪

**Comment by dideler** [comment]
Sounds like a good approach!

**Comment by 1natsu172** [comment]
@dideler, I'd like to discuss to you about something implementation. 🎙 

## Topic1: Re-discussion about display format.

@dideler said…

>In terms of format, we could do
><img width="176" alt="screen shot 2018-01-07 at 14 57 56" src="https://user-images.githubusercontent.com/497458/34650573-319d6286-f3bb-11e7-9cb5-e5194acb8898.png">
>or, as @1natsu172 originally suggested
><img width="187" alt="screen shot 2018-01-07 at 14 59 08" src="https://user-images.githubusercontent.com/497458/34650581-5a1817ce-f3bb-11e7-8952-523b154c735a.png">

We decided to make it the second option, due to grammatical problem of multilingual implementation. But I noticed when started implementing……we can easily implement the first option, if include DOM elements in "message.json". 💡 

> `_locales/en/message.json`
```json
{
  "extension_name": {
    "message": "Toggle YouTube Comments"
  },
  "extension_desc": {
    "message": "Hides YouTube comments by default"
  },
  "showCommentsText": {
    "message": "Show <span id=\"comments-count\">...</span> comments"
  },
  "hideCommentsText": {
    "message": "Hide comments"
  }
}
```

> `_locales/ja/message.json`
```json
{
  "extension_name": {
    "message": "Toggle YouTube Comments"
  },
  "extension_desc": {
    "message": "Hides YouTube comments by default"
  },
  "showCommentsText": {
    "message": "<span id=\"comments-count\">...</span>件のコメント"
  },
  "hideCommentsText": {
    "message": "コメントを閉じる"
  }
}
```

Maybe this method can be implemented with natural grammar even in multiple languages. How about that? 🤔 

By the way, when I looked at the YouTube user interface, I couldn't find the parenthesis notation like "(234, 521)". So the extension aims to blend the YouTube user interface, the first option may be better as well…?

## Topic2: Behavior of the number of comments when page navigation(page jump).

Whether or not to temporarily change the number of comments to "..." when page navigation.

But why,
Since YouTube is a single-page-application, DOM is not rewritten even if page navigation is done. So even if navigating to the next video, the number of comments of the previous video remains for a moment. It can be said that there is a lag before the extension rewrites to the number of comments for the next video.

#### 1st option.
This option will be slightly less code complexity.
But this may seem buggy for users due to the number of comments changing with lag. Moreover, it suddenly changes. The user may feel unnatural as though it had been rewritten by someone.
![111 2018-01-12 10 08 03-down](https://user-images.githubusercontent.com/7282145/34857629-24f8372c-f78f-11e7-9a9f-2d54504a28d1.png)

#### 2nd option.
This option will be slightly code complexity.
But this may be user-friendly. Because the number of comments for the previous video is not displayed on the next video.
![e 2018-01-12 10 08 03-down](https://user-images.githubusercontent.com/7282145/34857627-1a4d385e-f78f-11e7-9560-2d848906808c.png)

***

How do you feel? I want an opinion! 🙏

**Comment by dideler** [comment]
For topic 1, I still think the second option where we show the total count in parentheses is better. Because I think users could interpret the two formats differently.

> SHOW 100 COMMENTS

Can be interpreted as "show *all* 100 comments at once", but that's not what will happen.

> SHOW COMMENTS (100)

Can be interpreted as "show comments, and by the way, there are 100 total".

---

Another benefit is that the second option doesn't require code for pluralising "comments" because the count is not part of the sentence. E.g.

> SHOW 0 COMMENT
> SHOW 10 COMMENTS

vs

> SHOW COMMENTS (0)
> SHOW COMMENTS (10)

---

Related to topic 2, the second option will also be easier to read when the number of comments changes during navigation, because the text doesn't move around. E.g.

> SHOW 1,000 COMMENTS
> SHOW ... COMMENTS
> SHOW 543,210 COMMENTS

vs

> SHOW COMMENTS (1,000)
> SHOW COMMENTS
> SHOW COMMENTS (543,210)

_P.S. This example uses the user-friendly option from topic 2, which is the one I prefer._

---

In terms of the parentheses notation not blending in with the UI, there may be some other styles we can try. Maybe there is one that feels more natural. E.g.

> SHOW COMMENTS (100,00)
> SHOW COMMENTS: 100,00
> SHOW COMMENTS - 100,00
> SHOW COMMENTS  &nbsp; **100,00**
> ...

**Comment by 1natsu172** [comment]
@dideler thanks for your opinion! 😃 

## For topic 1

>>SHOW 100 COMMENTS

>Can be interpreted as "show all 100 comments at once", but that's not what will happen.

Certainly this is the same interpretation in Japanese.

>Another benefit is that the second option doesn't require code for pluralising "comments

This problem does not occur in Japanese grammar, but it occurs in English 😱 
Probably the same problem will occur in languages other than English.

I fully agree. Let's **continue with "option 2" for Topic 1.**

## For topic 2

> SHOW COMMENTS (1,000)
> SHOW COMMENTS
> SHOW COMMENTS (543,210)

I see! so let's also **adopt option 2 for topic 2.**

---

For in terms of the parentheses notation. I was watching again the YouTube UI where symbols are used with numbers. It was the only one in the feed of the top page.

> 695K views • 1 day ago

Following this …

> SHOW COMMENTS • 1,000
> SHOW COMMENTS
> SHOW COMMENTS • 543,210

I think this format. If extension uses a symbol for the number of comments.

Or most of Youtube's UI numbers are without symbol, so simple notation of space only ....

> SHOW COMMENTS&nbsp;&nbsp;1,000
> SHOW COMMENTS
> SHOW COMMENTS&nbsp;&nbsp;543,210

This may be perfectly personal preference issue ... 🤔 
I feel that the notation with "•" is easier to read. 👀

**Comment by dideler** [comment]
I think the '•' symbol won't serve exactly the same purpose if we were to use it.

> 695K views • 1 day ago

Here the '•' symbol is separating different types of content (views and days). But for us the content (comments and count) is closely related so it may be misleading to use the same separation symbol.

I like the whitespace notation, I'm noticing it a lot more on YouTube now that it's been brought to attention. But no matter which option we go with, we can always iterate and change it if it doesn't feel right or users don't like it 😄.

**Comment by 1natsu172** [comment]
It's exactly as you said, probably the '•' is separation symbol. And it's a fact that if users response is not good, we can change the continuously.

What we know now, symbols aren't frequent on YouTube UI and most of the numbers are white space notation…… For now, let's implement with only whitespace notation according to these 🚀  .

**Comment by dideler** [comment]
Resolved by da0fc01934562cc578ef418dcc6a9c8e1aedce76

**Comment by 1natsu172** [comment]
Thanks for your release work to the chrome store! 💯
